Unity-Technologies/ShaderGraph
# Vertex Position

![Image](https://image.ibb.co/gb7DE7/Screen_Shot_2018_04_13_at_11_29_15.png "Image")

# Vertex Modification

**REQUIRES PR: [LWRP #1193](https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/1193)**

## This PR
- Add `ShaderStageCapability` to all material slots
- Add `IsCompatibleStageWith` method to determine if two slots have compatible stages
- Add `ShaderStageCapability` to all mesh data requirement interfaces (needed for master node where different stages have different requirements)
- Add `ShaderStageExtensions`
- Add `includeIntermediateSpaces` check to `FromNodes<T>` (needed to exclude intermediate space translations from vertex and surface description input generation)
- Update `GetEffectiveShaderStage` and `GetEffectiveShaderStageCapability`to handle backwards edge detection (?)
- Add `Position` slot and `IMayRequirePosition` interface to both master nodes
- Add stage capability check to `IMayRequirePosition` and `IMayRequireNormal` on both master nodes
- Refactor master node shader generation to use `ShaderStringBuilder`
- Add `ShaderStageCapability` to `CodeFunctionNode` `SlotAttribute`
- Refactor `SurfaceMaterialOptions` to use `ShaderStringBuilder`
- Refactor `GetPropertiesDeclaration` (simpler)
- Refactor `ShaderGenerator` methods to use `ShaderStringBuilder`
- Major refactor to `GenerateStandardTransforms`
   - Add vertex requirements
   - Add generation of translations for vertex description
   - Add generation of vertex description inputs
   - Move everything to `ShaderStringBuilder`
   - Refactor code order for readability
   - Move all transformation code into a single block in output shader
- Refactor `GenerateSpaceTranslations` for vertex description use
- Refactor `GenerateSpaceTranslations` to use `ShaderStringBuilder`
- Refactor `GetPreviewSubShader` to use `ShaderStringBuilder`
- Refactor preview shader template for output readability
- Rename some `ShaderGeneratorNames` for consistency
- Add `BlockSemicolon` scope to `ShaderStringBuilder`
- Add indent level control to `IncreaseIndent` and `DecreaseIndent` in `ShaderStringBuilder`
- Add stage compability check to `SearchWindowProvider`
- Add initial framework for `Wireframe` type preview rendering
- Refactor `HDUnlitSubShader`
   - Move everything to `ShaderStringBuilder`
   - Refactor shader generation for output readability
   - Add `Alpha Threshold` slot and alpha clip functionality
   - Refactor C# layout to match LW subshaders
- Refactor `HDUnlitPassForward.template` to match changes to subshader
- Define various nodes as fragment stage only
   - Sample Texture 2D
   - Triplanar
   - Checkerboard
   - Create Normal
   - Ellipse
   - Rectangle
   - Rounded Rectangle
   - Polygon
